pymc/util.py

```diff
@@ -366,3 +366,12 @@ def wrapped(**kwargs):
         return core_function(**input_point)
 
     return wrapped
+
+
+def check_all_value_vars_in_model(vars, model):
+    notin = [v for v in vars if v not in model.value_vars]
+    if notin:
+        notin = list(map(get_var_name, notin))
+        raise ValueError(
+            "Some variables do not correspond to value variables in the model: " + str(notin)
```

Comment:
It might be worth explicitly recommending the use of `aesaraf.rvs_to_value_vars` here. I wasn't able to find that function until you told me about it!!

Reply:
Although that's the issue that motivated these changes I am not sure that would be the most common problem. I think passing variables from the wrong model is still the most likely reason to end up here, and in that case that advice would not be helpful at all. 

Given our sample size of 1, I would leave as is and reassess if more people complain about it

Reply:
lol fine.

Reply:
I agree, but the explicit message only makes sense when we are confident that's the most common cause. For instance, the error message we get when trying to iterate over a TensorVariable is more often that not misplaced. I never face it trying to call `sum` or `max`, but just when passing it into a function that expects an iterable.

```python
import aesara.tensor as at
x = at.vector("x")
for var in x:
  pass
```
```
TypeError: TensorType does not support iteration. Maybe you are using builtins.sum instead of aesara.tensor.math.sum? (Maybe .max?)
```

Back to the original suggestion. Are you convinced this is a common source of error for PyMC users?

Reply:
> that's the most common cause

That's the part that I disagree with! I'm not saying that you should replace this error with _only_ that suggestion. I'm just saying that this change is being made because of a specific issue that actually happens (in at least all models used by an entire field of researchers). This issue has a very simple fix that is deeply buried in a non-public interface. Seems worthwhile documenting that somewhere. Maybe this error message isn't the right place, but the current error message will seem very strange to those of us who _do_ pass in parameters that are a part of the model, just not one that you support!

Reply:
>  This issue has a very simple fix that is deeply buried in a non-public interface

I agree. I just don't think it's specific to `find_MAP` or even `check_all_value_variables_in_model` (but I could be convinced of the latter). I don't have a good alternative suggestion of where to put that information either...

Reply:
> I feel the error message is pretty clear now

I'd say this might be related to "value variables" meaning something specific to you. For those of us who are new to the new PyMC-specific terminology, I'd say that it's not super clear as written. I think that updating the `find_MAP` docs to specifically call out this new non-value-variable behavior and suggest `rvs_to_value_vars` as a possible solution, could be a great compromise!

Reply:
I'll try to update the docs to be more clear. 

Do note that `rvs_to_value_vars` is no longer sufficient, that gives you a graph built in terms of value variables, but you can still only pass input value variables, not any intermediate variable in the graph (such as a Deterministic). That's a breaking change from V3 that I am proposing to introduce in this PR because I think it's misleading to say we are optimizing with respect to such variables, when we are not.

To backtrack a bit, what is expected of a user when they call `find_MAP` (from now on) is that they specify a subset of free_RVs, not Deterministics, Potentials or other intermediate variables (and obviously not variables that don't belong to the model)

Behind the scenes we convert those to the respective value variables. We don't expect most users to ever have to worry about their existence (hence my hesitation to bring it up in a general error message). We do allow value variables to be passed directly, because this function may be called internally by other routines that already converted RVs to their values, but we don't expect users to be doing this directly.

The current error message makes more sense for developers than for users in that context. I'm happy to change it to be something like: `vars must contain only free_RVs or value variables`, or just `vars must contain only free_RVs` (after we checked that the input was not a value variable, to not make the message confusing in that case).

Reply:
Oh I see. I think it's a _huge_ mistake to remove the option of passing variables that aren't `free_RVs` - I'm now sad that I opened the initial issue! But I guess I'll just keep maintaining parallel packages that make PyMC usable in my community :D

Reply:
By the way, we didn't allow to pass those variables so far in V4 (I think it might have been allowed in V3), that's the "bug" you found where it was trying to optimize wrt to nothing. We were assuming users would pass `free_RVs` already.

The only reason I can see for a user to want to pass anything else is convenience + not knowing what model variables he is working with. In that case the solution as I described above is:

```python
def find_MAP_wrt_anything(graphs):
  # Replace RVs by values in whatever `vars` is
  value_graphs = pm.aesaraf.rvs_to_values(graphs)
  # Find the input value variables in those graphs
  input_values_graphs = pm.inputvars(value_graphs)
  return find_MAP(vars=input_values_graphs)
```

Reply:
While I agree with Ricardo that we are not really optimizing the deterministic, I think it is a reasonable "short cut" of expressing "optimized for all the dependent free_RVs of the deterministic".

How about a compromise that we raise a warning instead of an error if a deterministic from the same graph is inputted?

Reply:
Limited to `find_MAP` or also for every step, where we ask users which variables they want to assign to it?

Reply:
Limited to `find_MAP`, as you cannot assign sampler to underlying deterministic

Reply:
So do you suggest I replace the new error I added in `find_MAP`, by a nested try/except, where we also check if the variable corresponds to a deterministic in the model, and then issue the warning and use the underlying RVs? 

What if it's not a registered Deterministic?

Reply:
> Limited to find_MAP, as we cannot assign sampler to the RVs underlying deterministic

That exactly the same that's happening here in `find_MAP`. Except it's an optimizator and not a sampler.

Reply:
Sorry of not being more explicit, what i meant is (in `find_MAP`) something like:

```python
try:
    vars = get_value_vars_from_user_vars(vars, model)
except ValueError as exc:
    vars = pm.aesaraf.rvs_to_value_vars([vars])
    if vars is not Null:
          warnings.warn(
          "Intermediate variable (Deterministics or Potentials) was passed, "
          "find_MAP will optimized its dependent Free_RVs instead."
           )
    else:
        raise ValueError(...) from exc
```

I think this is ok to limit the behavior in find_MAP because find_MAP only works for all continues variable, while in sampler this is not guarantee. 

Let me know if I am missing something.

Reply:
Okay that sounds reasonable. Will update